generators/delimited/main.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+import sys
+import os
+import csv
+from time import strftime
+from unicode_csv import UnicodeDictReader
+
+def main():
+    usage = """Origins Delimited Generator.
+
+    Usage:
+        delimited <file_name> [--encoding=<encoding>] [--delimiter=<delimiter>] [--identifier=<indentifier>]
+
+    Options:
+        -h --help  Show usage info.
+        --encoding=<encoding>  The text encoding type of the file being read.
+        --delimiter=<delimiter>  The character used to separate items in the csv file.
+        --identifier=<identifier>  A comma-delimited (no spaces) list of fieldnames which make up a unique identifier for an entry in the file. e.g. --identifier=first,last,birthdate
+    """ #noqa
+
+    from docopt import docopt
+
+    # Trim 'origins generate' from the line.
+    argv = sys.argv[2:]
+    args = docopt(usage, argv=argv, version='0.1')
+
+   # Defaults
+    encoding = 'utf-8'
+    delimiter = ','
+    identifier = None
+
+    # Collect arg values
+    file_name = os.path.join(os.getcwd(), args['<file_name>'])
+    if args['--encoding'] is not None:
+        encoding = args['--encoding']
+
+    if args['--delimiter'] is not None:
+        delimiter = args['--delimiter']
+
+    if args['--identifier'] is not None:
+        identifier = args['--identifier'].split(",")
+
+    # No need to validate input because docopt makes sure all arguments are given
+    generate(file_name, encoding, delimiter, identifier)
+
+
+# Minimum possible header.
+HEADER = (
+    'entity',
+    'attribute',
+    'value',
+)
+
+def generate(file_name, encoding, delimiter, identifier):
+    # Creating a CSV writer that feeds to stdout.
+    w = csv.writer(sys.stdout)
+
+    # Write the header to start.
+    w.writerow(HEADER)
+
+    # Create each fact
+    with open(file_name, 'rU') as f:
+        sniff = 1024
+        dialect = None
+
+        # Infer various properties about the file
+        # Sample the file to determine the dialect
+        sample = '\n'.join([l for l in f.readlines(sniff)])
+        f.seek(0)
+
+        # Determine dialect
+        sniffer = csv.Sniffer()
+        dialect = sniffer.sniff(sample)
+
+        reader = UnicodeDictReader(f, dialect=dialect, delimiter=delimiter, encoding=encoding)
+
+        for line in reader:
+            # Assemble the identifier's fieldnames into a single string for output.
+            id = ''
+            if identifier is not None:
+                id = build_id(line, identifier)
+            else:
+                id = line[reader.fieldnames[0]]
```

Comment:
`reader.fieldnames[0]` does not change across iterations, it can be set outside of the loop:

```python
reader = ...

if not identifier:
    # Default to the first column.
    identifier = reader.fieldnames[:1]

for line in reader:
    id = build_id(line, identifier)
    ...
```